backward_diff discretization of cont. transfer function

zolabar opened the pull request:
# Added backward difference discretization method to physics.control.lti

#### References to other Issues or PRs

It is a 1 to 1 extension of [PR 24558](https://github.com/sympy/sympy/pull/24558)


#### Brief description of what is fixed or changed

Added function

```
backward_diff(tf, sample_per)
```

In full analogy of 

```
bilinear(tf, sample_per)
```

In practical use of contiuous transfer function $H(s)$  (corresponding to differential eq.)one eventually needs the dicsretized version $H(z)$ (corresponding to difference eq.).

* bilinear renders coefficients for trapezoidal integration scheme
* backward_diff renders coefficients for backward difference integration scheme 

Both methdos preserve the stability of the continuous input transfer function.

#### Other comments

SymPy would obtain more than one symbolic discretizatin methods of continuous transfer functions.

As the numerical counterparts from the scientific computing with python eco system

https://python-control.readthedocs.io/en/latest/generated/control.sample_system.html

https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.cont2discrete.html#scipy.signal.cont2discrete

#### Release Notes


<!-- BEGIN RELEASE NOTES -->

NO ENTRY

<!-- END RELEASE NOTES -->

zolabar committed:
backward_diff

zolabar commented:
@hanspi42, I added 

backward_diff(tf, sample_per)

I think that it is practical for implementation usage.

I made use of you mathematical tricks and your code prototype.

In 

https://github.com/sympy/sympy/discussions/24643

I opened a discussion on partial fraction decomposition method and impulse invariant discretization method, may be those ideas could be discussed in that discussion.

zolabar committed:
Update __init__.py

zolabar committed:
.. autofunction:: backward_diff

zolabar committed:
backward_diff in example

asmeurer commented on sympy/physics/control/lti.py:
```diff
@@ -76,6 +77,54 @@ def bilinear(tf, sample_per):
+        as [a, b], [c, d].
+
+        Examples
+
```
Remove the blank line here.

sylee957 replied:
```suggestion
```

asmeurer commented on sympy/physics/control/lti.py:
```diff
@@ -76,6 +77,54 @@ def bilinear(tf, sample_per):
+        [L + R*T, -L]
+        """
+
+        z = Symbol('z') # discrete variable z
```
This should be passed in as a parameter to the function.

zolabar replied:
@asmeurer , thank you for feedback.  Should z be an Input parameter, even if it is used only as dummy variable, i.e. If it is not part of the output?

zolabar replied:
@asmeurer , @hanspi42, 

in `bilinear` and in `backward_diff` the dummy discrete variable "z" is now the internal var symbol, i.e. the variable of the corresponding transfer function.

Then, there is no neeed anymore of defining a new Symbol (z = Symbol('z'))

Since `bilinear` and in `backward_diff` return only coefficients this OK from my point of view.

zolabar replied:
@asmeurer , @hanspi42, can I resolve this conversation? Thank you

sylee957 commented on sympy/physics/control/lti.py:
```diff
@@ -76,6 +77,54 @@ def bilinear(tf, sample_per):
+        discretized with the backward difference transform method.
+        H(z) is obtained from the continuous transfer function H(s)
+        by substituting s(z) =  (z-1)/(T*z) into H(s), where T is the
+        sample period.
+        H(z) corresponds to the difference equation of the by backward difference scheme.
+        Coefficients are falling, i.e. H(z) = (az+b)/(cz+d) is returned
+        as [a, b], [c, d].
+
+        Examples
+
+        ========
+
+        >>> from sympy.physics.control.lti import TransferFunction, backward_diff
+        >>> from sympy.abc import s, L, R, T
+        >>> tf = TransferFunction(1, s*L + R, s)
+        >>> numZ, denZ = backward_diff(tf, T)
+        >>> numZ
+        [T, 0]
+        >>> denZ
+        [L + R*T, -L]
+        """
+
+        z = Symbol('z') # discrete variable z
+        T = sample_per  # and sample period T
+        s = tf.var
+
+        np = tf.num.as_poly(s).all_coeffs()
+        dp = tf.den.as_poly(s).all_coeffs()
+
+        # The next line results from multiplying H(z) with z^N/z^N
+
+        N = max(len(np), len(dp)) - 1
+        num = Add(*[ T**(N-i)*c*(z-1)**i*(z)**(N-i) for c, i in zip(np[::-1], range(len(np))) ])
+        den = Add(*[ T**(N-i)*c*(z-1)**i*(z)**(N-i) for c, i in zip(dp[::-1], range(len(dp))) ])
+
+        num_coefs = num.as_poly(z).all_coeffs()
+        den_coefs = den.as_poly(z).all_coeffs()
+
+
+        return num_coefs, den_coefs
```
The function shouldn't be doubly indented.

zolabar replied:
👍🏽

zolabar committed:
Update sympy/physics/control/lti.py

Co-authored-by: S.Y. Lee  <sylee957@gmail.com>

faze-geek commented on sympy/physics/control/lti.py:
```diff
@@ -76,6 +77,53 @@ def bilinear(tf, sample_per):
+        H(z) is obtained from the continuous transfer function H(s)
+        by substituting s(z) =  (z-1)/(T*z) into H(s), where T is the
+        sample period.
+        H(z) corresponds to the difference equation of the by backward difference scheme.
```
`of the by backward difference scheme` doesn't seem to make sense. Should it simply be `of the backward difference scheme` ?

zolabar replied:
👍🏽

zolabar replied:
> `of the by backward difference scheme` doesn't seem to make sense. Should it simply be `of the backward difference scheme` ?

-> corrected

zolabar committed:
corrected docstring of backward_diff

zolabar committed:
dummy discrete variable can be internal var symbol

in bilinear and in backward_diff the dummy discrete variable "z" can be internal var symbol, i.e. the variable of the corresponding transfer function.

Then, there is no neeed of defining a new Symbol (z = Symbol('z'))

zolabar committed:
strip_whitespace

hanspi42 commented:
The code looks OK for me, but the documentation is a bit weak. You now write in two functions:

```
H(z) corresponds to the difference equation of the trapezoidal scheme.
```

Either you should omit this, or write down the difference equation.

zolabar commented:
> The code looks OK for me, but the documentation is a bit weak. You now write in two functions:
> 
> ```
> H(z) corresponds to the difference equation of the trapezoidal scheme.
> ```
> 
> Either you should omit this, or write down the difference equation.

@hanspi42 , thank you very much for the Assessment!

I don‘t see the documentation bug in the current version…

https://output.circle-artifacts.com/output/job/f709aeca-5990-4b84-a8c9-2b5bd9c44135/artifacts/0/doc/_build/html/modules/physics/control/lti.html#sympy.physics.control.lti.backward_diff

Did I oversee something?

hanspi42 commented:
> https://output.circle-artifacts.com/output/job/f709aeca-5990-4b84-a8c9-2b5bd9c44135/artifacts/0/doc/_build/html/modules/physics/control/lti.html#sympy.physics.control.lti.backward_diff
> 
> Did I oversee something?

Well, there are no difference equatios shown there, it is just mentioned that difference equations exist. It is not helpful in the documentation if you refer to a difference equation that you don't show. You should show it, or not mention it.

zolabar commented:
> > https://output.circle-artifacts.com/output/job/f709aeca-5990-4b84-a8c9-2b5bd9c44135/artifacts/0/doc/_build/html/modules/physics/control/lti.html#sympy.physics.control.lti.backward_diff
> > Did I oversee something?
> 
> Well, there are no difference equatios shown there, it is just mentioned that difference equations exist. It is not helpful in the documentation if you refer to a difference equation that you don't show. You should show it, or not mention it.

@hanspi42 , ok, thanks. I see,  I‘ll choose one of those two suggested options!

zolabar committed:
Removed ref to diff eqs

Introducing general difference equations not possible in few lines.

Adding examples without the concept also not clear

zolabar commented:
> > > https://output.circle-artifacts.com/output/job/f709aeca-5990-4b84-a8c9-2b5bd9c44135/artifacts/0/doc/_build/html/modules/physics/control/lti.html#sympy.physics.control.lti.backward_diff
> > > Did I oversee something?
> > 
> > 
> > Well, there are no difference equatios shown there, it is just mentioned that difference equations exist. It is not helpful in the documentation if you refer to a difference equation that you don't show. You should show it, or not mention it.
> 
> @hanspi42 , ok, thanks. I see, I‘ll choose one of those two suggested options!

@hanspi42 , I removed the references on the diference equations

hanspi42 commented:
All is OK, thank you for your contribution!

zolabar commented:
> All is OK, thank you for your contribution!

👍
